linked_list/linked_list.py

```diff
@@ -84,26 +152,47 @@ def visit(self):
         
         print(", ".join(helper_list))
 
+
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        prev = None
+        current = self.head
+        while(current is not None):
+            next = current.next
+            current.next = prev
+            prev = current
+            current = next
+        self.head = prev
+
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        pointer_1 = self.head
+        pointer_2 = self.head
+        while pointer_2 and pointer_2.next:
+            pointer_1 = pointer_1.next
+            pointer_2 = pointer_2.next.next
+        return pointer_1.value
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):

```

Comment:
I don't get how this recursion works because you're sending `self.head.next` into the place of `self` in the method parameters...